File: identity-agent-core/m63/cesr.go
package m63

import (
	"encoding/base64"
	"fmt"
)

const (
	CipherSuiteIAHybrid1   = "IA-HYBRID-1"
	CESRMLDSA65Verkey      = "1PDA"
	CESRMLDSA65Sig         = "1PDS"
	CESRMLKEM768Encap      = "1PKM"
	CESRX25519Pubkey       = "1PXB"
	MLDSA65VerkeyBytes     = 1952
	MLDSA65SigBytes        = 3309
	MLKEM768EncapBytes     = 1184
	X25519PubkeyBytes      = 32
	Ed25519PubkeyBytes     = 32
	Ed25519SigBytes        = 64
	CtrControllerIdxSigs   = "-A"
	C2Message              = "m63-c2-hybrid-signature-golden-vector"
)

var (
	C2Ed25519Seed = func() []byte {
		out := make([]byte, 32)
		for i := range out {
			out[i] = byte((i + 0x21) % 256)
		}
		return out
	}()
	C2MLDSASeed = []byte("m63-c2-hybrid-signature-golden!!")
)

// keri Base64 index alphabet (matches keripy intToB64).
var b64Chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_"

func intToB64(i int, length int) string {
	d := make([]byte, 0, length)
	for length > 0 && (len(d) < length || i > 0) {
		d = append([]byte{b64Chars[i%64]}, d...)
		i /= 64
	}
	for len(d) < length {
		d = append([]byte{'A'}, d...)
	}
	return string(d)
}

func b64ToInt(s string) int {
	n := 0
	for _, c := range s {
		n = n*64 + int(bytesIndexRune(b64Chars, c))
	}
	return n
}

func bytesIndexRune(s string, r rune) int {
	for i, c := range s {
		if c == r {
			return i
		}
	}
	return 0
}

// MatterFixedQB64 encodes fixed-size CESR material (keri 1.1.17 Matter._infil).
func MatterFixedQB64(code string, raw []byte) (string, error) {
	if len(code) < 1 {
		return "", fmt.Errorf("matter code required")
	}
	ps := (3 - (len(raw) % 3)) % 3
	cs := len(code)
	padded := make([]byte, ps+len(raw))
	copy(padded[ps:], raw)
	b64 := base64.RawURLEncoding.EncodeToString(padded)
	if cs%4 > len(b64) {
		return "", fmt.Errorf("invalid matter encoding for code %q", code)
	}
	return code + b64[cs%4:], nil
}

func EncodeLargeFixed(code string, raw []byte, expectedLen int) (string, error) {
	if len(code) != 4 {
		return "", fmt.Errorf("provisional CESR code must be 4 chars, got %q", code)
	}
	if expectedLen > 0 && len(raw) != expectedLen {
		return "", fmt.Errorf("expected %d raw bytes for %s, got %d", expectedLen, code, len(raw))
	}
	return code + base64.RawURLEncoding.EncodeToString(raw), nil
}

File: identity_agent_ui/rust/src/pqc/cesr.rs
//! Provisional IA-HYBRID-1 CESR codes (C3-pinned wire bytes in golden_vectors.json).

pub const CIPHER_SUITE_IA_HYBRID_1: &str = "IA-HYBRID-1";
pub const CESR_MLDSA65_VERKEY: &str = "1PDA";
pub const CESR_MLDSA65_SIG: &str = "1PDS";
pub const CESR_MLKEM768_ENCAP: &str = "1PKM";
pub const CESR_X25519_PUBKEY: &str = "1PXB";
pub const MLDSA65_VERKEY_BYTES: usize = 1952;
pub const MLDSA65_SIG_BYTES: usize = 3309;
pub const ED25519_SIG_BYTES: usize = 64;
pub const CTR_CONTROLLER_IDX_SIGS: &str = "-A";
pub const C2_MESSAGE: &str = "m63-c2-hybrid-signature-golden-vector";

pub fn c2_ed25519_seed() -> [u8; 32] {
    std::array::from_fn(|i| ((i + 0x21) % 256) as u8)
}

pub fn c2_mldsa_seed() -> [u8; 32] {
    *b"m63-c2-hybrid-signature-golden!!"
}

const B64_CHARS: &[u8] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";

pub fn int_to_b64(i: usize, length: usize) -> String {
    let mut n = i;
    let mut d: Vec<u8> = Vec::new();
    while length > 0 && (d.len() < length || n > 0) {
        d.insert(0, B64_CHARS[n % 64]);
        n /= 64;
    }
    while d.len() < length {
        d.insert(0, b'A');
    }
    String::from_utf8(d).unwrap_or_default()
}

pub fn b64_to_int(s: &str) -> usize {
    s.bytes().fold(0usize, |n, c| {
        let idx = B64_CHARS.iter().position(|&x| x == c).unwrap_or(0);
        n * 64 + idx
    })
}
pub const MLKEM768_ENCAP_BYTES: usize = 1184;
pub const X25519_PUBKEY_BYTES: usize = 32;
pub const ED25519_PUBKEY_BYTES: usize = 32;

/// Fixed-size CESR Matter encoding (keri 1.1.17 Matter._infil).
pub fn matter_fixed_qb64(code: &str, raw: &[u8]) -> Result<String, String> {
    if code.is_empty() {
        return Err("matter code required".into());
    }
    let ps = (3 - (raw.len() % 3)) % 3;
    let cs = code.len();
    let mut padded = vec![0u8; ps + raw.len()];
    padded[ps..].copy_from_slice(raw);
    let b64 = base64::encode_config(&padded, base64::URL_SAFE_NO_PAD);
    if cs % 4 > b64.len() {
        return Err(format!("invalid matter encoding for code {code}"));
    }
    Ok(format!("{}{}", code, &b64[cs % 4..]))
}

pub fn encode_large_fixed(code: &str, raw: &[u8], expected_len: usize) -> Result<String, String> {
    if code.len() != 4 {
        return Err(format!("provisional CESR code must be 4 chars, got {code:?}"));
    }
    if expected_len > 0 && raw.len() != expected_len {
        return Err(format!(
            "expected {expected_len} raw bytes for {code}, got {}",
            raw.len()
        ));
    }
    Ok(format!(
        "{}{}",
        code,
        base64::encode_config(raw, base64::URL_SAFE_NO_PAD)
    ))
}

pub fn ed25519_verfer_qb64(raw32: &[u8]) -> Result<String, String> {
    if raw32.len() != ED25519_PUBKEY_BYTES {
        return Err("Ed25519 public key must be 32 bytes".into());
    }
    matter_fixed_qb64("D", raw32)
}

pub fn blake3_qb64(data: &[u8]) -> Result<String, String> {
    let sum = blake3::hash(data);
    matter_fixed_qb64("E", sum.as_bytes())
}